src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/WritableBufferWriter.cs

```diff
@@ -0,0 +1,360 @@
+﻿using System;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Internal.Infrastructure
+{
+    public struct WritableBufferWriter
+    {
+        private const int _maxULongByteLength = 20;
+
+        [ThreadStatic]
+        private static byte[] _numericBytesScratch;
+
+        private int _index;
+        private WritableBuffer _buffer;
+        private int _length;
+        private Span<byte> _span;
+
+        public WritableBufferWriter(WritableBuffer buffer)
+        {
+            _buffer = buffer;
+            _index = 0;
+            _span = buffer.Buffer.Span;
+            _length = _span.Length;
+        }
+
+        public void WriteFast(byte[] source)
+        {
+            WriteFast(source, 0, source.Length);
+        }
+
+        public void WriteFast(ArraySegment<byte> source)
+        {
+            WriteFast(source.Array, source.Offset, source.Count);
+        }
+
+        public void WriteFast(byte[] source, int offset, int length)
+        {
+            var dest = _span;
+            var index = _index;
+            var destLength = _length;
+            var sourceLength = length;
+
+            if (sourceLength <= destLength)
+            {
+                ref byte pSource = ref source[offset];
+                ref byte pDest = ref dest[index];
+                Unsafe.CopyBlockUnaligned(ref pDest, ref pSource, (uint)sourceLength);
+                Advance(sourceLength);
+                return;
+            }
+
+            WriteMultiBuffer(source, offset, length);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void Grow()
+        {
+            // Advance the buffer to what we've written for this block
+            _buffer.Advance(_index);
+            _buffer.Ensure();
+            _index = 0;
+            _span = _buffer.Buffer.Span;
+            _length = _span.Length;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void Advance(int length)
+        {
+            _index += length;
+            _length -= length;
+        }
+
+        public void Commit()
```

Comment:
Ugh, I was thinking about this but tried to avoid. Would require having 2 `Commit` calls and it was pretty fast without this optimization